test/Grpc.AspNetCore.Server.Tests/PercentEncodingHelpersTests.cs

```diff
@@ -0,0 +1,84 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Linq;
+using Grpc.AspNetCore.Server.Internal;
+using NUnit.Framework;
+
+namespace Grpc.AspNetCore.Server.Tests
+{
+    [TestFixture]
+    public class PercentEncodingHelpersTests
+    {
+        [TestCaseSource(nameof(ValidPercentEncodingTestCases))]
+        public void EncodeMessageTrailer(string value, string expectedEncodedValue)
+        {
+            // Arrange & Act
+            var encodedValue = PercentEncodingHelpers.PercentEncode(value);
+
+            // Assert
+            Assert.AreEqual(expectedEncodedValue, encodedValue);
+        }
+
+        [TestCaseSource(nameof(ValidPercentEncodingTestCases))]
+        public void DecodeMessageTrailer(string expectedDecodedValue, string value)
+        {
+            // Arrange & Act
+            var decodedValue = Uri.UnescapeDataString(value);
+
+            // Assert
+            Assert.AreEqual(expectedDecodedValue, decodedValue);
+        }
+
+        [TestCase("%", "%")]
+        [TestCase("%A", "%A")]
+        [TestCase("%A%", "%A%")]
+        [TestCase("%AG", "%AG")]
+        [TestCase("\0", "\0")]
+        public void DecodeInvalidMessageTrailer(string expectedDecodedValue, string value)
+        {
+            // Arrange & Act
+            var decodedValue = Uri.UnescapeDataString(value);
+
+            // Assert
+            Assert.AreEqual(expectedDecodedValue, decodedValue);
+        }
+
+        private static TestCaseData[] ValidPercentEncodingTestCases =
```

Comment:
Some of the java test cases might come handy
https://github.com/grpc/grpc-java/blob/80c3c992a66aa21ccf3e12e38000316e45f97e64/api/src/test/java/io/grpc/StatusTest.java#L65

Reply:
Ok, I'll add them.